src/ekat/ekat_assert.hpp

```diff
@@ -25,16 +26,24 @@
 #define EKAT_BACKTRACE __FILE__ << ":" << __LINE__
 #endif
 
-// Internal do not call directly
-#define IMPL_THROW(condition, msg, exception_type)    \
-  do {                                                \
-    if ( ! (condition) ) {                            \
-      std::stringstream _ss_;                         \
-      _ss_ << "\n FAIL:\n" << #condition  << "\n";   \
-      _ss_ << EKAT_BACKTRACE;                         \
-      _ss_ << "\n" << msg;                            \
-      throw exception_type(_ss_.str());               \
-    }                                                 \
+// Internal do not call directly.
+// NOTE: the ... at the end is to allow using EKAT_REQUIRE with
+//       a variadic number of args, adding placeholders at the end
+//       to ensure that the call to IMPL_THROW matches the signature
+#define IMPL_THROW(condition, msg, exception_type, ...)                           \
+  do {                                                                            \
+    if ( ! (condition) ) {                                                        \
+      std::stringstream _ss_;                                                     \
+      _ss_ << "\n FAIL:\n" << #condition  << "\n";                                \
+      _ss_ << EKAT_BACKTRACE;                                                     \
+      _ss_ << "\n" << msg;                                                        \
+      if constexpr (std::is_constructible<exception_type, std::string>::value) {  \
+        throw exception_type(_ss_.str());                                         \
+      } else {                                                                    \
+        std::cerr << _ss_;                                                        \
```

Comment:
This prints the message to screen before raising the exception.

This case allows to handle most (if not all) std library exception types. User defined exceptions MUST fit one of these two constructors (empty or taking a `std::string` or `const char*`) in order for code to compile.